torchao/prototype/mx_formats/nvfp4_tensor.py

```diff
@@ -0,0 +1,619 @@
+        sliced_data = aten.slice.Tensor(x._data, dim, start, end, step)
+    elif dim == 1:
+        # Slicing along reduction dim - must align with block boundaries
+        if start is not None:
+            assert start % x._block_size == 0, (
+                f"Start index {start} must be a multiple of block_size {x._block_size}"
+            )
+
+        if end is not None:
+            assert end % x._block_size == 0, (
+                f"End index {end} must be a multiple of block_size {x._block_size}"
+            )
+
+        sliced_data = aten.slice.Tensor(x._data, dim, start, end, step)
+
+        # Calculate which scale blocks to keep
+        start_block = 0 if start is None else start // x._block_size
+        end_block = None if end is None else end // x._block_size
+
+        # Slice the scale tensor accordingly
+        sliced_scale = aten.slice.Tensor(scale_shaped, 1, start_block, end_block, step)
+    else:
+        raise ValueError(
+            f"NVFP4Tensor only supports slicing along dimensions 0 and 1, got dim={dim}"
+        )
+
+    return NVFP4Tensor(
+        sliced_scale,
+        x._per_tensor_scale,  # Unchanged per-tensor scale
+        sliced_data,
+        x._block_size,
+        x._orig_dtype,
+        x.mm_config,
+    )
+
+
+@implements([aten.t.default])
+def nvfp4_t(func, types, args, kwargs):
+    # For now, only transpose(input, 0, 1) is supported.
+    old = args[0]
+    new = NVFP4Tensor(
+        old._scale_e4m3,
+        old._per_tensor_scale,
+        old._data.t(),
+        old._block_size,
+        old._orig_dtype,
+        old.mm_config,
+    )
+    return new
+
+
+@implements([aten.view.default])
+def nvfp4_view_op(func, types, args, kwargs):
+    data = args[0]._data
+    new_size = args[1]
+    new_size = tensor_size_hp_to_fp4x2(new_size, data.is_contiguous())
+    new_data = func(data, new_size, *args[2:], **kwargs)
+    return NVFP4Tensor(
+        args[0]._scale_e4m3,
+        args[0]._per_tensor_scale,
+        new_data,
+        args[0]._block_size,
+        args[0]._orig_dtype,
+        args[0].mm_config,
+    )
+
+
+def _addmm_nvfp4_dispatch(
+    a: NVFP4Tensor, b: NVFP4Tensor, aten_op, bias: Optional[torch.Tensor] = None
+) -> torch.Tensor:
+    """
+    Core implementation shared between nvfp4_mm, nvfp4_addmm, and nvfp4_linear.
+    The only difference is whether bias is None or not.
+    """
+    assert a._data.is_contiguous()
+    assert b._data.t().is_contiguous()
+    assert a._block_size == 16, f"NVFP4 requires block_size=16, got {a._block_size}"
+    assert b._block_size == 16, f"NVFP4 requires block_size=16, got {b._block_size}"
+
+    M, K = a.shape[0], a.shape[1]
+    N = b.shape[1]
+
+    # Swizzle Dizzle
+    a_scale = a._scale_e4m3.view(M, K // a._block_size)
+    b_scale = b._scale_e4m3.view(N, K // b._block_size)
+    a_scale_blocked = to_blocked(a_scale)
+    b_scale_blocked = to_blocked(b_scale)
```

Comment:
I have wondered about this for MX-dtype as well. It makes sense for MX-dtype to have scale swizzling here since we may want the layout to be vendor-neutral. But NVFP4 is specific to NVIDIA, so why not put this under `to_nvfp4()`?

Reply:
This is a good question, for mx I recently confirmed that AMD does not require swizziling, this is a good point about NVFP4, I am actually going update this code path to 	`cache` the swizzled layout for the weight